File: backend/app/models/travel_request.py
"""
Domain model — represents a travel request document as stored in MongoDB.

This is NOT a Beanie/ODM model. It is a plain Pydantic model that describes
the shape of a MongoDB document, used internally by the repository and
service layers.  Motor handles all database I/O.
"""
from __future__ import annotations

from datetime import date, datetime, timezone
from enum import Enum
from typing import Any

from pydantic import BaseModel, Field


# ── Enumerations ──────────────────────────────────────────────────────────────


class TripType(str, Enum):
    """Allowed trip categories."""

    VACATION = "Vacation"
    BUSINESS = "Business"
    ADVENTURE = "Adventure"


class BudgetRange(str, Enum):
    """Allowed budget brackets."""

    LOW = "Low"
    MEDIUM = "Medium"
    HIGH = "High"


class TravelRequestStatus(str, Enum):
    """Lifecycle status of a travel request."""

    PENDING = "Pending"
    APPROVED = "Approved"
    REJECTED = "Rejected"
    CLOSED = "Closed"

class ApprovalStatus(str, Enum):
    """Approval workflow status."""

    NOT_REQUIRED = "Not Required"
    PENDING = "Pending"
    APPROVED = "Approved"
    REJECTED = "Rejected"


class TaskStatus(str, Enum):
    """Status of a fulfillment task."""

    PENDING = "Pending"
    COMPLETED = "Completed"


class FulfillmentTask(BaseModel):
    """A single fulfillment task attached to an approved travel request."""

    id: str
    title: str
    description: str
    status: TaskStatus = TaskStatus.PENDING
    completed_at: datetime | None = None


class Approval(BaseModel):
    """
    Approval workflow information.
    """

    required: bool
    status: ApprovalStatus
    approver: str | None = None
    approved_at: datetime | None = None
    rejected_at: datetime | None = None
    remarks: str | None = None
    

# ── Domain model ──────────────────────────────────────────────────────────────


class TravelRequestModel(BaseModel):
    """
    Full representation of a travel-request MongoDB document.

    Field naming convention
    -----------------------
    - ``id`` is the string-serialised ``_id`` (ObjectId) value.
      The repository is responsible for the ObjectId ↔ str conversion.
    - All datetime values are UTC-aware.

    Placeholder fields
    ------------------
    ``weather``, ``recommendation``, ``approval`` and ``tasks`` are
    reserved for future phases.  They are stored as ``None`` / ``[]``
    on creation so that the document schema is stable and downstream
    features can populate them without a schema migration.
    """

    # ── Identity ──────────────────────────────────────────────────────────────
    id: str = Field(..., description="String-encoded MongoDB ObjectId (_id).")

    # ── User-supplied fields ──────────────────────────────────────────────────
    destination_city: str = Field(
        ...,
        min_length=1,
        max_length=100,
        description="Name of the destination city.",
    )
    travel_date: date = Field(..., description="Planned departure date (ISO 8601).")
    trip_type: TripType = Field(..., description="Category of the trip.")
    budget_range: BudgetRange = Field(..., description="Approximate budget bracket.")
    special_needs: bool = Field(
        default=False,
        description="Whether the traveller has special requirements.",
    )
    notes: str | None = Field(
        default=None,
        max_length=1000,
        description="Required when special_needs is True; optional otherwise.",
    )

    # ── System-managed fields ─────────────────────────────────────────────────
    status: TravelRequestStatus = Field(
        default=TravelRequestStatus.PENDING,
        description="Lifecycle status managed by the approval workflow.",
    )

    # ── Future-phase placeholders ─────────────────────────────────────────────
    weather: dict[str, Any] | None = Field(
        default=None,
        description="[Phase 3] Weather data fetched for the destination and date.",
    )
    recommendation: dict[str, Any] | None = Field(
        default=None,
        description="[Phase 4] AI recommendation output.",
    )
    approval: Approval | None = Field(
        default=None,
        description="[Phase 5] Approval workflow information.",
    )
    tasks: list[dict[str, Any]] = Field(
        default_factory=list,
        description="[Phase 6] Checklist / task items attached to this request.",
    )

    # ── Timestamps ────────────────────────────────────────────────────────────
    created_at: datetime = Field(
        default_factory=lambda: datetime.now(tz=timezone.utc),
        description="UTC timestamp of document creation.",
    )
    updated_at: datetime = Field(
        default_factory=lambda: datetime.now(tz=timezone.utc),
        description="UTC timestamp of the last update.",
    )

    model_config = {
        # Allow population from MongoDB dicts that use native Python types.
        "populate_by_name": True,
    }

File: frontend/src/types/travel-request.ts
export const TripType = {
  VACATION: "Vacation",
  BUSINESS: "Business",
  ADVENTURE: "Adventure",
} as const;
export type TripType = typeof TripType[keyof typeof TripType];

export const BudgetRange = {
  LOW: "Low",
  MEDIUM: "Medium",
  HIGH: "High",
} as const;
export type BudgetRange = typeof BudgetRange[keyof typeof BudgetRange];

export const TravelRequestStatus = {
  PENDING: "Pending",
  APPROVED: "Approved",
  REJECTED: "Rejected",
  CLOSED: "Closed",
} as const;
export type TravelRequestStatus = typeof TravelRequestStatus[keyof typeof TravelRequestStatus];

export const ApprovalStatus = {
  NOT_REQUIRED: "Not Required",
  PENDING: "Pending",
  APPROVED: "Approved",
  REJECTED: "Rejected",
} as const;
export type ApprovalStatus = typeof ApprovalStatus[keyof typeof ApprovalStatus];

export interface Approval {
  required: boolean;
  status: ApprovalStatus;
  approver?: string | null;
  approved_at?: string | null;
  rejected_at?: string | null;
  remarks?: string | null;
}

export interface WeatherLocation {
  name: string;
  latitude: number;
  longitude: number;
  country: string | null;
}

export interface WeatherForecast {
  date: string;
  temperature_max: number;
  temperature_min: number;
  precipitation_probability: number;
  weather_code: number;
  weather_description: string;
}

export interface WeatherSummary {
  location: WeatherLocation;
  forecast: WeatherForecast;
}

export interface Recommendation {
  suitable: boolean;
  title: string;
  message: string;
  risk_level: "low" | "medium" | "high";
}

export interface TravelRequest {
  id: string;
  destination_city: string;
  travel_date: string;
  trip_type: TripType;
  budget_range: BudgetRange;
  special_needs: boolean;
  notes: string | null;
  status: TravelRequestStatus;
  weather?: WeatherSummary | null;
  recommendation?: Recommendation | null;
  approval?: Approval | null;
  tasks: FulfillmentTask[];
  created_at: string;
  updated_at: string;
}

export interface CreateTravelRequestRequest {
  destination_city: string;
  travel_date: string;
  trip_type: TripType;
  budget_range: BudgetRange;
  special_needs?: boolean;
  notes?: string | null;
}

export interface UpdateTravelRequestRequest {
  destination_city?: string;
  travel_date?: string;
  trip_type?: TripType;
  budget_range?: BudgetRange;
  special_needs?: boolean;
  notes?: string | null;
}

export interface TravelRequestListResponse {
  total: number;
  skip: number;
  limit: number;
  data: TravelRequest[];
}

export const TaskStatus = {
  PENDING: "Pending",
  COMPLETED: "Completed",
} as const;
export type TaskStatus = typeof TaskStatus[keyof typeof TaskStatus];

export interface FulfillmentTask {
  id: string;
  title: string;
  description: string;
  status: TaskStatus;
  completed_at: string | null;
}
